### ISSUE TITLE
pypi package / travis deploys

### ISSUE THREAD

**Issue by mikkokotila** [Esukhia/pybo#7]
Hey Guys, really great to see that things are moving :)  For pypi it would be good idea to deploy Travis-ci first and then add in the end the deploy automatically to pypi when certain criteria is fulfilled (for example a merge from master to production branch or something like that). Let me know when is a good time to do some testing, and I'll do a PR with the travis.yml based on that. I can try to include some very basic getting started guide  in the PR. What do you think?

**Comment by mikkokotila** [comment]
Just to add about travis-ci in case you guys are not familiar with it from other projects, it helps take away some of the biggest headaches that come with the kind of development pace as you have. Also it adds zero complexity as code is not effected, it's just one file, and runs the test scripts before anything gets merged. Generally it's a good idea to block merges to master unless Travis tests pass. With a well thought out test script, Travis helps to completely eradicate...

<img src='https://camo.githubusercontent.com/99795876091b147750b439140434bcbc0acc2ed1/68747470733a2f2f732d6d656469612d63616368652d616b302e70696e696d672e636f6d2f6f726967696e616c732f38332f66372f38652f38336637386536326665623935616363383564303030616166363335306432332e6a7067' width=300>

...and all the suffering that comes with it.

**Comment by mikkokotila** [comment]
That can be set based on your preference. You will have to set it up from the settings of the repo, so you have full control once travis.yml is in place. Every PR/commit will automatically run travis-ci, and then based on the settings for a given branch in github, there might be a merge block. Admins can override merge blocks (sometimes that might be handy but generally not a good practice). The automatic deploy to pypi can be either branch specific (e.g. just production) or tag specific. One nice feature in travis is that many python versions can be tested simultaneously. Also code coverage, such as coveralls, is just 2 lines of code extra (because the test script is there already). IMO language code should have very high code coverage. We can talk about testing first, I'll just include some simple functional tests for now. In the longer term functional tests should be avoided, with preference for validity tests. Code coverage can also be used as a criteria to block merge, but that's slightly trickier as once the coverage is already high, it becomes increasingly hard to always keep the same coverage or increase it with each merge. But I do think that for the longer term, it should be incorporated as it forces a situation where new merged code always has 100% coverage, which in turn forces the committer to actually test their code (actually testing would automatically yield a code that results in 100% coverage when included in the test script). This in turn introduces something really powerful; contributors think very carefully how they write their code. Not something crucial at this point, but as soon as pybo is used for some other production such as Tibet Editor, that seems to be the way to go. 

Talking about settings, unless you already did it, it's good to have a delete block on the master branch. That's from the same part of the settings where the test passing related blocks are set.

**Comment by drupchen** [comment]
You are right about code coverage. I would be the first one to be refused a merge, I fear. hehehe
Frankly speaking, I don't think I would have the time to code if I had to have 100% coverage all the time...

What you describe really looks promising. I am eager to see it in action!

**Comment by mikkokotila** [comment]
Can you create a @dev or some other branch than master, and I'll do a PR to that. Given how good this is looking already, I recommend that you stop making commits directly to master. Generally if you move fast, it's good to have something like daily-dev which you keep merging to dev daily, and then you keep merging master to dev (!) weekly and then maybe monthly sync everything with master. The master -> dev approach tends to avoid many headaches where you have 2-3 contributors with their own branches. The github conflict resolution really sucks and actually has some bugs, so it's very good if that can be entirely avoided. Which is little funny given that conflict resolution is the core functionality that everything else is built around...

In the meantime can you sync your repo with http://travis-ci.com/ and https://coveralls.io/. Both should be simply a matter of: 

1) login in to the service using your github account
2) going to your github settings to authorize the service 
3) switching on the repo from the service (e.g. travis) 

That should do it for both. I'll do the rest in the PR. Then once I make the PR, the tests will run automatically and we should see the results in the merge page. 

By the way, do you support 2.7 as well or only 3? It should be mostly backward compatible, but some things like zip() for example return a different kind of object so that might break it.

**Comment by mikkokotila** [comment]
ok yea, it fails with 2.7 but passes on 3.4, 3.5 and 3.6 on travis with Ubuntu Trusty. 

I'll do some testing + test script improvements in a bit.

**Comment by drupchen** [comment]
I agree for the need to have a dev branch. I had one, but then deleted it. 
Also, the purpose for which I started pybo altogether is now met: it finally has all the functionalities required to be functional in the tibetaneditor and now fully replaces pytib.

Pytib has two parts: an initial segmentation that took care of separating affixed particles and a second step to adjust the segmentation using rules that needed to be manually written.

pybo does the first part using pybotrie and splitaffixed, the second one using adjusttokens.
The added functionality is that the rules can be obtained by using RDRPOSTagger and adapting the rules to fit those used by adjusttokens. This way, adjusting the initial segmentation can be improved by training RDR on a larger corpus of well segmented corpus.

Now that this is done, the focus will be on proper documentation, improving tests and debugging. In short, everything that is not directly the core of the project but is necessary for it to be useful and used. So in the forseeable future, there won't be any drastical changes in the project. Also, I will have much less time from now onwards, so I think you come at the right moment for everything you are proposing.

I have created a develop branch and I will have to wait until next week to create the travis and coveralls accounts. I will let you know when this is done. Thanks again for your help!
